src/wallet/wallet.cpp

```diff
@@ -3225,26 +3225,19 @@ bool CWallet::GetOutpointAndKeysFromOutput(const COutput& out, COutPoint& outpoi
 int CWallet::CountInputsWithAmount(CAmount nInputAmount)
 {
     CAmount nTotal = 0;
-    {
-        LOCK2(cs_main, cs_wallet);
-        for (std::map<uint256, CWalletTx>::const_iterator it = mapWallet.begin(); it != mapWallet.end(); ++it)
-        {
-            const CWalletTx* pcoin = &(*it).second;
-            if (pcoin->IsTrusted()){
-                int nDepth = pcoin->GetDepthInMainChain();
 
-                for (unsigned int i = 0; i < pcoin->tx->vout.size(); i++) {
-                    COutput out = COutput(pcoin, i, nDepth, true, true, false);
-                    COutPoint outpoint = COutPoint(out.tx->GetHash(), out.i);
+    LOCK2(cs_main, cs_wallet);
+
+    for (const auto& outpoint : setWalletUTXO) {
+        const auto it = mapWallet.find(outpoint.hash);
+        if (it == mapWallet.end() || it->second.GetDepthInMainChain() < 0) continue;
 
-                    if(out.tx->tx->vout[out.i].nValue != nInputAmount) continue;
-                    if(!CPrivateSend::IsDenominatedAmount(pcoin->tx->vout[i].nValue)) continue;
-                    if(IsSpent(out.tx->GetHash(), i) || IsMine(pcoin->tx->vout[i]) != ISMINE_SPENDABLE || !IsDenominated(outpoint)) continue;
+        const CTxOut &txout = it->second.tx->vout[outpoint.n];
 
-                    nTotal++;
-                }
-            }
-        }
+        if (txout.nValue != nInputAmount) continue;
+        if (!CPrivateSend::IsDenominatedAmount(txout.nValue)) continue;
```

Comment:
Does this really make sense? The methods looks for inputs with `nInputAmount`, so to have the same effect it should be enough to just check `IsDenominatedAmount(nInputAmount)` before the loop starts...but then, why is this limitation here at all? The method says `CountInputsWithAmount` and not `CountDenominatedInputs` ;) The only caller of this method looks like it would be fine with removing this check here.